### PR TITLE
G-API: Mosaic handle corner cases

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by TolyaTalamanov** [comment]
@smirnov-alexey @AsyaPronina

File: modules/gapi/src/api/render_ocv.cpp [hunk 1]
```diff
@@ -16,16 +16,27 @@ namespace draw
...
     for(int i = 0; i < crop_y; i += cellSz ) {
         for(int j = 0; j < crop_x; j += cellSz) {
-            auto cell_roi = msc_roi(cv::Rect(j, i, cellSz, cellSz));
+            cell_roi = msc_roi(cv::Rect(j, i, cellSz, cellSz));
             cell_roi = cv::mean(cell_roi);
         }
     }
+
+    // Handle tail
+    cell_roi = msc_roi(cv::Rect(crop_x, 0, msc_roi.cols - crop_x, crop_y));
```
Comment (alalek) [inline modules/gapi/src/api/render_ocv.cpp]: - Proper handling of "no tail" cases
- Are there two rectangles only for the tail processing?
- I believe you should not use rectangles greater than `cellSz x cellSz`. `0 .. crop_y` violates this condition.

Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_ocv.cpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@alalek Could you look at it again ?

File: modules/gapi/test/render/gapi_render_tests_ocv.cpp [hunk 2]
```diff
@@ -493,15 +493,41 @@ INSTANTIATE_TEST_CASE_P(RenderNV12OCVTestFTextsImpl, RenderNV12OCVTestFTexts,
                             Values(cv::Scalar(0, 255, 0))));
 #endif // HAVE_FREETYPE
 
+// FIXME Implement a macros to instantiate the tests because BGR and NV12 have the same parameters
+
 INSTANTIATE_TEST_CASE_P(RenderBGROCVTestMosaicsImpl, RenderBGROCVTestMosaics,
                         Combine(Values(cv::Size(1280, 720)),
```
Comment (alalek) [inline modules/gapi/test/render/gapi_render_tests_ocv.cpp]: @dmatveev Consider avoiding using of HD-sized images in accuracy tests.

These tests are filtered from:
- Debug pre-commit build
- coverage
- valgrind

Only VGA and less sized images are run for these builder types.

---

BTW, If it is highly required to work with HD resolution, then don't make them parametrized.

File: modules/gapi/src/api/render_ocv.cpp (hunk 1 above)
Comment (alalek) [reply modules/gapi/src/api/render_ocv.cpp]: > Proper handling of "no tail" cases

Test still crashes in "Debug" build due extra checks (don't look on pre-commit, it doesn't care about 720p test cases).

File: modules/gapi/test/render/gapi_render_tests_ocv.cpp (hunk 2 above)
Comment (alalek) [inline modules/gapi/test/render/gapi_render_tests_ocv.cpp]: @dmatveev Consider avoiding using of HD-sized images in accuracy tests of G-API.

These tests are filtered from:
- Debug pre-commit build
- coverage runs
- valgrind runs

Only VGA and less sized images are run for these builder types.

---

BTW, If it is highly required to work with HD+ resolution, then don't make these tests parametrized on image resolution.

Comment (dmatveev) [reply modules/gapi/test/render/gapi_render_tests_ocv.cpp]: ok we'll address it in our next tests refactoring round, thanks!

**Comment by dmatveev** [comment]
@alalek This stability thing is required for the upcoming internal release, can we have it in 4.2 (which I believe will serve as a basis for that)?

**Comment by alalek** [comment]
Current patch is buggy.
It throws exceptions in Debug builds (related tests are filtered out on public CI due HD resolution - try run them on local machine).

**Comment by TolyaTalamanov** [comment]
@alalek Could you take a look ?

**Comment by dmatveev** [comment]
Thanks!

**Comment by TolyaTalamanov** [comment]
Great! Thanks.

**Pull request merged**